# Which Zipcode in King County Has the Most Expensive Sales Price in 2021 on  Average?

## Introduction

This project uses the dataset "kc_house_data.csv" from Kaggle.com. It includes sales information of homes sold between May 2014 and May 2015 in King County, including Seattle. The sales information contains price of sales and 19 features (i.e. number of bedroom, view) on 21613 observations. 

To answer the question, Y variable is homes' price of sales in U.S. dollars and the feature- zipcode is the first X variable that is used to compare home prices. In addition, to build the price prediction model and to investigate what features make houses in certain zipcodes more expensive, 3 additional features have been chosen to be X variables- square footage of the homes' interior living space(sqft_living), quality of construction and design(grade), and view. Note that grade is an index from 1 to 13, where 1-3 falls short of building construction and design, 7 is an average level, and 11-13 indicates a high quality level. View is an index from 0 to 4 of how good the view of the property was. The first two features are chosen based on their relatively strong correlation with home's sales price, and view is chosen because King County owns great views that zipcodes with better views may be more popular than other zipcodes.

## Import Data 

In [1]:
# ! pip install qeds
# import libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import qeds
%matplotlib inline

# activate plot theme
qeds.themes.mpl_style();

from IPython.display import display

In [2]:
# import and display dataset in dataframe format
data_path = "/Users/cindyxin/Documents/ECO225/ECO225Project/Data/kc_house_data.csv"
data = pd.read_csv(data_path)
data

id             date     price  bedrooms  bathrooms  \
0      7129300520  20141013T000000  221900.0         3       1.00   
1      6414100192  20141209T000000  538000.0         3       2.25   
2      5631500400  20150225T000000  180000.0         2       1.00   
3      2487200875  20141209T000000  604000.0         4       3.00   
4      1954400510  20150218T000000  510000.0         3       2.00   
...           ...              ...       ...       ...        ...   
21608   263000018  20140521T000000  360000.0         3       2.50   
21609  6600060120  20150223T000000  400000.0         4       2.50   
21610  1523300141  20140623T000000  402101.0         2       0.75   
21611   291310100  20150116T000000  400000.0         3       2.50   
21612  1523300157  20141015T000000  325000.0         2       0.75   

       sqft_living  sqft_lot  floors  waterfront  view  ...  grade  \
0             1180      5650     1.0           0     0  ...      7   
1             2570      7242     2.0           0     0  ...      7   
2              770     10000     1.0           0     0  ...      6   
3             1960      5000     1.0           0     0  ...      7   
4             1680      8080     1.0           0     0  ...      8   
...            ...       ...     ...         ...   ...  ...    ...   
21608         1530      1131     3.0           0     0  ...      8   
21609         2310      5813     2.0           0     0  ...      8   
21610         1020      1350     2.0           0     0  ...      7   
21611         1600      2388     2.0           0     0  ...      8   
21612         1020      1076     2.0           0     0  ...      7   

       sqft_above  sqft_basement  yr_built  yr_renovated  zipcode      lat  \
0            1180              0      1955             0    98178  47.5112   
1            2170            400      1951          1991    98125  47.7210   
2             770              0      1933             0    98028  47.7379   
3            1050            910      1965             0    98136  47.5208   
4            1680              0      1987             0    98074  47.6168   
...           ...            ...       ...           ...      ...      ...   
21608        1530              0      2009             0    98103  47.6993   
21609        2310              0      2014             0    98146  47.5107   
21610        1020              0      2009             0    98144  47.5944   
21611        1600              0      2004             0    98027  47.5345   
21612        1020              0      2008             0    98144  47.5941   

          long  sqft_living15  sqft_lot15  
0     -122.257           1340        5650  
1     -122.319           1690        7639  
2     -122.233           2720        8062  
3     -122.393           1360        5000  
4     -122.045           1800        7503  
...        ...            ...         ...  
21608 -122.346           1530        1509  
21609 -122.362           1830        7200  
21610 -122.299           1020        2007  
21611 -122.069           1410        1287  
21612 -122.299           1020        1357  

[21613 rows x 21 columns]

In [3]:
# information of variables (check for null and type)
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21613 entries, 0 to 21612
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             21613 non-null  int64  
 1   date           21613 non-null  object 
 2   price          21613 non-null  float64
 3   bedrooms       21613 non-null  int64  
 4   bathrooms      21613 non-null  float64
 5   sqft_living    21613 non-null  int64  
 6   sqft_lot       21613 non-null  int64  
 7   floors         21613 non-null  float64
 8   waterfront     21613 non-null  int64  
 9   view           21613 non-null  int64  
 10  condition      21613 non-null  int64  
 11  grade          21613 non-null  int64  
 12  sqft_above     21613 non-null  int64  
 13  sqft_basement  21613 non-null  int64  
 14  yr_built       21613 non-null  int64  
 15  yr_renovated   21613 non-null  int64  
 16  zipcode        21613 non-null  int64  
 17  lat            21613 non-null  float64
 18  long  

In [4]:
# read all columns
pd.set_option("display.max_columns",21)
data

id             date     price  bedrooms  bathrooms  \
0      7129300520  20141013T000000  221900.0         3       1.00   
1      6414100192  20141209T000000  538000.0         3       2.25   
2      5631500400  20150225T000000  180000.0         2       1.00   
3      2487200875  20141209T000000  604000.0         4       3.00   
4      1954400510  20150218T000000  510000.0         3       2.00   
...           ...              ...       ...       ...        ...   
21608   263000018  20140521T000000  360000.0         3       2.50   
21609  6600060120  20150223T000000  400000.0         4       2.50   
21610  1523300141  20140623T000000  402101.0         2       0.75   
21611   291310100  20150116T000000  400000.0         3       2.50   
21612  1523300157  20141015T000000  325000.0         2       0.75   

       sqft_living  sqft_lot  floors  waterfront  view  condition  grade  \
0             1180      5650     1.0           0     0          3      7   
1             2570      7242     2.0           0     0          3      7   
2              770     10000     1.0           0     0          3      6   
3             1960      5000     1.0           0     0          5      7   
4             1680      8080     1.0           0     0          3      8   
...            ...       ...     ...         ...   ...        ...    ...   
21608         1530      1131     3.0           0     0          3      8   
21609         2310      5813     2.0           0     0          3      8   
21610         1020      1350     2.0           0     0          3      7   
21611         1600      2388     2.0           0     0          3      8   
21612         1020      1076     2.0           0     0          3      7   

       sqft_above  sqft_basement  yr_built  yr_renovated  zipcode      lat  \
0            1180              0      1955             0    98178  47.5112   
1            2170            400      1951          1991    98125  47.7210   
2             770              0      1933             0    98028  47.7379   
3            1050            910      1965             0    98136  47.5208   
4            1680              0      1987             0    98074  47.6168   
...           ...            ...       ...           ...      ...      ...   
21608        1530              0      2009             0    98103  47.6993   
21609        2310              0      2014             0    98146  47.5107   
21610        1020              0      2009             0    98144  47.5944   
21611        1600              0      2004             0    98027  47.5345   
21612        1020              0      2008             0    98144  47.5941   

          long  sqft_living15  sqft_lot15  
0     -122.257           1340        5650  
1     -122.319           1690        7639  
2     -122.233           2720        8062  
3     -122.393           1360        5000  
4     -122.045           1800        7503  
...        ...            ...         ...  
21608 -122.346           1530        1509  
21609 -122.362           1830        7200  
21610 -122.299           1020        2007  
21611 -122.069           1410        1287  
21612 -122.299           1020        1357  

[21613 rows x 21 columns]

## Data Cleaning

In [5]:
# make a copy of data to make changes on
cleaned_data = data.copy()
cleaned_data

id             date     price  bedrooms  bathrooms  \
0      7129300520  20141013T000000  221900.0         3       1.00   
1      6414100192  20141209T000000  538000.0         3       2.25   
2      5631500400  20150225T000000  180000.0         2       1.00   
3      2487200875  20141209T000000  604000.0         4       3.00   
4      1954400510  20150218T000000  510000.0         3       2.00   
...           ...              ...       ...       ...        ...   
21608   263000018  20140521T000000  360000.0         3       2.50   
21609  6600060120  20150223T000000  400000.0         4       2.50   
21610  1523300141  20140623T000000  402101.0         2       0.75   
21611   291310100  20150116T000000  400000.0         3       2.50   
21612  1523300157  20141015T000000  325000.0         2       0.75   

       sqft_living  sqft_lot  floors  waterfront  view  condition  grade  \
0             1180      5650     1.0           0     0          3      7   
1             2570      7242     2.0           0     0          3      7   
2              770     10000     1.0           0     0          3      6   
3             1960      5000     1.0           0     0          5      7   
4             1680      8080     1.0           0     0          3      8   
...            ...       ...     ...         ...   ...        ...    ...   
21608         1530      1131     3.0           0     0          3      8   
21609         2310      5813     2.0           0     0          3      8   
21610         1020      1350     2.0           0     0          3      7   
21611         1600      2388     2.0           0     0          3      8   
21612         1020      1076     2.0           0     0          3      7   

       sqft_above  sqft_basement  yr_built  yr_renovated  zipcode      lat  \
0            1180              0      1955             0    98178  47.5112   
1            2170            400      1951          1991    98125  47.7210   
2             770              0      1933             0    98028  47.7379   
3            1050            910      1965             0    98136  47.5208   
4            1680              0      1987             0    98074  47.6168   
...           ...            ...       ...           ...      ...      ...   
21608        1530              0      2009             0    98103  47.6993   
21609        2310              0      2014             0    98146  47.5107   
21610        1020              0      2009             0    98144  47.5944   
21611        1600              0      2004             0    98027  47.5345   
21612        1020              0      2008             0    98144  47.5941   

          long  sqft_living15  sqft_lot15  
0     -122.257           1340        5650  
1     -122.319           1690        7639  
2     -122.233           2720        8062  
3     -122.393           1360        5000  
4     -122.045           1800        7503  
...        ...            ...         ...  
21608 -122.346           1530        1509  
21609 -122.362           1830        7200  
21610 -122.299           1020        2007  
21611 -122.069           1410        1287  
21612 -122.299           1020        1357  

[21613 rows x 21 columns]

In [6]:
# create sales yr in a new column using sales date
from datetime import datetime

cleaned_data["datetime"] = pd.to_datetime(cleaned_data.date)
cleaned_data["yr_sales"] = cleaned_data["datetime"].dt.year

# create renovation dummy for houses renovated
cleaned_data["renovated"] = cleaned_data["yr_renovated"].apply(lambda x : 0 if x == 0 else 1)

In [7]:
# drop id, date, zipcode, and yr_renovated
cleaned_data = cleaned_data.drop(["date", "datetime", "id", "yr_renovated"], axis = 1)

In [8]:
# quick read over dataset 
cleaned_data.head()

price  bedrooms  bathrooms  sqft_living  sqft_lot  floors  waterfront  \
0  221900.0         3       1.00         1180      5650     1.0           0   
1  538000.0         3       2.25         2570      7242     2.0           0   
2  180000.0         2       1.00          770     10000     1.0           0   
3  604000.0         4       3.00         1960      5000     1.0           0   
4  510000.0         3       2.00         1680      8080     1.0           0   

   view  condition  grade  sqft_above  sqft_basement  yr_built  zipcode  \
0     0          3      7        1180              0      1955    98178   
1     0          3      7        2170            400      1951    98125   
2     0          3      6         770              0      1933    98028   
3     0          5      7        1050            910      1965    98136   
4     0          3      8        1680              0      1987    98074   

       lat     long  sqft_living15  sqft_lot15  yr_sales  renovated  
0  47.5112 -122.257           1340        5650      2014          0  
1  47.7210 -122.319           1690        7639      2014          1  
2  47.7379 -122.233           2720        8062      2015          0  
3  47.5208 -122.393           1360        5000      2014          0  
4  47.6168 -122.045           1800        7503      2015          0

In [9]:
# information of variables (check for null and type)
cleaned_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21613 entries, 0 to 21612
Data columns (total 20 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   price          21613 non-null  float64
 1   bedrooms       21613 non-null  int64  
 2   bathrooms      21613 non-null  float64
 3   sqft_living    21613 non-null  int64  
 4   sqft_lot       21613 non-null  int64  
 5   floors         21613 non-null  float64
 6   waterfront     21613 non-null  int64  
 7   view           21613 non-null  int64  
 8   condition      21613 non-null  int64  
 9   grade          21613 non-null  int64  
 10  sqft_above     21613 non-null  int64  
 11  sqft_basement  21613 non-null  int64  
 12  yr_built       21613 non-null  int64  
 13  zipcode        21613 non-null  int64  
 14  lat            21613 non-null  float64
 15  long           21613 non-null  float64
 16  sqft_living15  21613 non-null  int64  
 17  sqft_lot15     21613 non-null  int64  
 18  yr_sal

## Choosing X Variables by Observing Correlation of House Price and Individual Features

In [10]:
# correlation with y variable
corr_matrix = cleaned_data.corr()
corr_matrix['price'].sort_values(ascending = False)

price            1.000000
sqft_living      0.702035
grade            0.667434
sqft_above       0.605567
sqft_living15    0.585379
bathrooms        0.525138
view             0.397293
sqft_basement    0.323816
bedrooms         0.308350
lat              0.307003
waterfront       0.266369
floors           0.256794
renovated        0.126092
sqft_lot         0.089661
sqft_lot15       0.082447
yr_built         0.054012
condition        0.036362
long             0.021626
yr_sales         0.003576
zipcode         -0.053203
Name: price, dtype: float64

Observing the correlation of each feature and price (Y variable), we are going to choose features that are most correlated to price as X variables. These features are sqft_living, and grade. In addition, view is also really important in a sense because there are great views in King County that zipcodes with great views may be more popular than zipcodes without great views. Thus, view would also be a X variable. 

In [11]:
# new dataframe with chosen X variables and Y variables
cleaned_data1 = data[["zipcode", "price", "sqft_living", "grade", "view"]]
cleaned_data1

zipcode     price  sqft_living  grade  view
0        98178  221900.0         1180      7     0
1        98125  538000.0         2570      7     0
2        98028  180000.0          770      6     0
3        98136  604000.0         1960      7     0
4        98074  510000.0         1680      8     0
...        ...       ...          ...    ...   ...
21608    98103  360000.0         1530      8     0
21609    98146  400000.0         2310      8     0
21610    98144  402101.0         1020      7     0
21611    98027  400000.0         1600      8     0
21612    98144  325000.0         1020      7     0

[21613 rows x 5 columns]

## Summary Statistics

In [12]:
# summary statistics of chosen Xs and Y
cleaned_data1.describe()

zipcode         price   sqft_living         grade          view
count  21613.000000  2.161300e+04  21613.000000  21613.000000  21613.000000
mean   98077.939805  5.400881e+05   2079.899736      7.656873      0.234303
std       53.505026  3.671272e+05    918.440897      1.175459      0.766318
min    98001.000000  7.500000e+04    290.000000      1.000000      0.000000
25%    98033.000000  3.219500e+05   1427.000000      7.000000      0.000000
50%    98065.000000  4.500000e+05   1910.000000      7.000000      0.000000
75%    98118.000000  6.450000e+05   2550.000000      8.000000      0.000000
max    98199.000000  7.700000e+06  13540.000000     13.000000      4.000000

The summary statistics shows that price of houses in 2014-2015 ranged from 75,000 to 7,700,000 U.S.D.  

In [13]:
# dataset group by zipcode 
group = cleaned_data1.groupby("zipcode")

for zipcode, zip_data in group:
    print(zipcode)
    print(zip_data)

98001
       zipcode     price  sqft_living  grade  view
38       98001  240000.0         1220      7     0
67       98001  287000.0         2240      7     0
114      98001  480000.0         3230      9     3
213      98001  465000.0         2714      9     0
292      98001  360000.0         2160      9     0
...        ...       ...          ...    ...   ...
21391    98001  269950.0         2244      7     0
21479    98001  312891.0         2300      8     0
21511    98001  480000.0         2940      9     0
21565    98001  341780.0         2260      7     0
21598    98001  541800.0         3118      9     2

[362 rows x 5 columns]
98002
       zipcode     price  sqft_living  grade  view
18       98002  189000.0         1200      7     0
25       98002  233000.0         1710      6     0
243      98002  225000.0         1300      6     0
248      98002  250000.0         2010      7     0
576      98002  266000.0         1805      7     0
...        ...       ...          ...    ...  

[105 rows x 5 columns]
98103
       zipcode     price  sqft_living  grade  view
17       98103  485000.0         1600      7     0
111      98103  570000.0         1260      6     0
116      98103  518500.0         1590      8     0
128      98103  822500.0         2320      7     0
149      98103  511000.0         1430      7     0
...        ...       ...          ...    ...   ...
21567    98103  690000.0         1540      9     0
21574    98103  610950.0         1680      8     0
21592    98103  475000.0         1190      8     0
21595    98103  520000.0         1530      8     0
21608    98103  360000.0         1530      8     0

[602 rows x 5 columns]
98105
       zipcode      price  sqft_living  grade  view
54       98105   920000.0         2730      8     0
55       98105   885000.0         2830      9     0
211      98105   625504.0         1270      8     0
308      98105   550700.0         1010      6     0
330      98105   880000.0         2800      9     0
...        ...   

In [14]:
# summary statistics of Xs and Y grouped by zipcode
group.describe()

price                                                              \
         count          mean            std       min       25%        50%   
zipcode                                                                      
98001    362.0  2.808047e+05   98979.230092  100000.0  215000.0   260000.0   
98002    199.0  2.342840e+05   52041.448659   95000.0  199975.0   235000.0   
98003    280.0  2.941113e+05  107597.662236  128000.0  229000.0   267475.0   
98004    317.0  1.355927e+06  745820.089407  425000.0  845000.0  1150000.0   
98005    168.0  8.101649e+05  268753.720249  400000.0  637775.0   765475.0   
...        ...           ...            ...       ...       ...        ...   
98177    255.0  6.761854e+05  435849.885258  245560.0  428500.0   554000.0   
98178    262.0  3.106128e+05  161044.746227   90000.0  223250.0   278277.0   
98188    136.0  2.890783e+05  103106.579944  110000.0  228750.0   264000.0   
98198    280.0  3.028789e+05  154799.271902   95000.0  216375.0   265000.0   
98199    317.0  7.918208e+05  397868.234000  323000.0  535000.0   689800.0   

                              sqft_living               ... grade        \
               75%        max       count         mean  ...   75%   max   
zipcode                                                 ...               
98001     319375.0   850000.0       362.0  1900.856354  ...   8.0  10.0   
98002     268500.0   389000.0       199.0  1627.743719  ...   7.0   8.0   
98003     330000.0   950000.0       280.0  1928.882143  ...   8.0  11.0   
98004    1691000.0  7062500.0       317.0  2909.022082  ...  10.0  13.0   
98005     911250.0  1960000.0       168.0  2656.803571  ...   9.0  12.0   
...            ...        ...         ...          ...  ...   ...   ...   
98177     739475.0  3800000.0       255.0  2323.333333  ...   8.0  13.0   
98178     349792.5  1700000.0       262.0  1729.351145  ...   7.0  10.0   
98188     335000.0   707000.0       136.0  1802.772059  ...   7.0  10.0   
98198     340000.0  1350000.0       280.0  1745.360714  ...   7.0  11.0   
98199     950000.0  3200000.0       317.0  2161.798107  ...   9.0  13.0   

          view                                               
         count      mean       std  min  25%  50%  75%  max  
zipcode                                                      
98001    362.0  0.102210  0.529512  0.0  0.0  0.0  0.0  4.0  
98002    199.0  0.010050  0.141776  0.0  0.0  0.0  0.0  2.0  
98003    280.0  0.214286  0.653479  0.0  0.0  0.0  0.0  3.0  
98004    317.0  0.305994  0.840657  0.0  0.0  0.0  0.0  4.0  
98005    168.0  0.095238  0.538773  0.0  0.0  0.0  0.0  4.0  
...        ...       ...       ...  ...  ...  ...  ...  ...  
98177    255.0  0.815686  1.349157  0.0  0.0  0.0  2.0  4.0  
98178    262.0  0.534351  1.109331  0.0  0.0  0.0  0.0  4.0  
98188    136.0  0.147059  0.602826  0.0  0.0  0.0  0.0  3.0  
98198    280.0  0.592857  1.172544  0.0  0.0  0.0  0.0  4.0  
98199    317.0  0.552050  1.019418  0.0  0.0  0.0  1.0  3.0  

[70 rows x 32 columns]

In [15]:
# 2014-2015 average sales price of homes in different zipcodes.
zipcode_mean_price = group["price"].mean()
zipcode_price_rank = zipcode_mean_price.sort_values(ascending = False)
zipcode_price_rank.reset_index()

zipcode         price
0     98039  2.160607e+06
1     98004  1.355927e+06
2     98040  1.194230e+06
3     98112  1.095499e+06
4     98102  9.012583e+05
..      ...           ...
65    98148  2.849086e+05
66    98001  2.808047e+05
67    98032  2.512962e+05
68    98168  2.403284e+05
69    98002  2.342840e+05

[70 rows x 2 columns]

In 2014-2015, homes in zipcode 98039 have the most expensive sales price on average. This area is located by Lake Washington. 

In [16]:
# 2014-2015 average grade on quality of construction and design of homes in different zipcodes.
zipcode_mean_grade = group["grade"].mean()
zipcode_grade_rank = zipcode_mean_grade.sort_values(ascending = False)
zipcode_grade_rank.reset_index()

zipcode     grade
0     98039  9.560000
1     98075  9.008357
2     98040  8.960993
3     98006  8.795181
4     98077  8.782828
..      ...       ...
65    98106  6.865672
66    98146  6.822917
67    98178  6.816794
68    98002  6.693467
69    98168  6.509294

[70 rows x 2 columns]

In 2014-2015, homes in zipcode 98039 have the best average grade on quality of construction and design of homes. Note that it's the same zipcode that has on average the most expensive sales price in this time period.

In [17]:
# 2014-2015 average grade on view of homes in different zipcodes.
zipcode_mean_view = group["view"].mean()
zipcode_view_rank = zipcode_mean_view.sort_values(ascending = False)
zipcode_view_rank.reset_index()

zipcode      view
0     98070  1.084746
1     98177  0.815686
2     98040  0.783688
3     98166  0.665354
4     98006  0.634538
..      ...       ...
65    98168  0.014870
66    98077  0.010101
67    98002  0.010050
68    98031  0.007299
69    98148  0.000000

[70 rows x 2 columns]

In 2014-2015, homes in zipcode 98070 have the best average grade on homes' view. However, this is not one of the several most expensive zipcodes shown in zipcode_price_rank. In general, the top and bottom zipcodes in this table barely appears on zipcode_price_rank. This indicates a relatively weak correlation of view and price of home.

In [18]:
# 2014-2015 average square footage of the apartments' interior living space in different zipcodes.
zipcode_mean_sqft_living = group["sqft_living"].mean()
zipcode_sqft_living_rank = zipcode_mean_sqft_living.sort_values(ascending = False)
zipcode_sqft_living_rank.reset_index()

zipcode  sqft_living
0     98039  3800.900000
1     98040  3106.833333
2     98075  3016.370474
3     98004  2909.022082
4     98006  2888.295181
..      ...          ...
65    98148  1610.526316
66    98107  1579.548872
67    98126  1543.596045
68    98106  1487.883582
69    98168  1468.624535

[70 rows x 2 columns]

In 2014-2015, homes in zipcode 98039 have the biggest square footage of the homes' interior living space. At the same time, this zipcode also has the best average grade on quality of construction and design, and the zipcode is on average the most expensive zipcode in time period.